### PR TITLE
chore: RFC for handling discarded events

### PR DESCRIPTION
**Pull request by bruceg**
[Rendered](https://github.com/vectordotdev/vector/blob/bruceg/discarded-events-rfc/rfcs/2022-08-25-12217-handling-discarded-events.md)

Ref: #1772 #12217 #13549

### CODE REVIEW

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+
+In order to avoid breaking existing configurations, this validation will initially produce a
+deprecation warning, notifying the user that an output is present but unhandled. In a later version,
+this deprecation will be turned into a hard error, preventing the acceptance of a configuration with
+unhandled outputs.
+
+We will also add a new command-line option and environment variable to opt into the stricter
+validation for users that want the additional assurance this provides.
+
+#### Simplify Discarded Output Handling
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:73]: I find the names used here awkward, but admittedly don't have any suggestions.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+
+This enforced handling will create considerable complication for most users, as well as the
+potential of increased overhead and reduced performance if extra components have to be configured
+just to handle the outputs that now need to be connected somewhere. To reduce this overhead, we will
+add a new optional configuration setting to indicate the internal disposition of each output:
+`outputs.NAME.disposition`. This will have two non-default values, `"drop"` to mark all events going
+to that output as delivered and then drop them, and `"reject"` to mark them as having failed.  If
+this setting is not present and a component has an unhandled output, a default behavior of `"drop"`
+will be assumed, which matches the current behavior.
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:82]: Will this default be changing when the previously mentioned deprecation warning becomes an error? If not, doesn't this invalidate the benefit of that check?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+buffers would be a good approach, but that path towards that is less clear than solving the problem
+in isolation.
+
+## Outstanding Questions
+
+- Do we want to automate the process of rewriting configurations that have missing output handlers,
+  or can we just leave the behavior as producing an error?
+
+- Should the option for turning unhandled output enforcement into a hard error be written as a more
+  generic switch to turn _all_ deprecations into errors?
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:336]: Yes! 👍

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+
+- Do we need separate outputs to distinguish between discards and errors?
+
+- Do we want to have a third value for `output.X.disposition` to allow for marking events as either
+  permanently failed (ie rejected) vs temporarily failed (ie errored)?
+
+- What should be done with the existing discarded event metrics? Should they always be emitted, or
+  only when the output isn't consumed by another component? Do we need another disposition marker to
+  indicate discards are not to be counted as errors?
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:345]: 🤔 if we're forcing it to be handled, I would think they could be counted like we do with other event streams - events/bytes sent/received. Practically, that seems like a hard sell given the work we've just put into the discarded metrics and compatibility with the ui.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 2 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:82]: The default will be removed at that point. IOW the behavior default is what will be deprecated.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+- Incompatible event type
+- Unhandled metric value type
+- Unwired component output
+- Failure to send to next component
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:38]: Is this type of failure possible in cases other than a non-graceful shutdown (i.e. panic)?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+Vector has a number of ways in which components may drop events due that are outside of the
+operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+- Incompatible event type
+- Unhandled metric value type
+- Unwired component output
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:37]: These three probably aren't entirely outside of the operator's control.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+1. Add a new standardized output to all components that may discard events.
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
+
+We will introduce a new output to all components that would otherwise discard events, named
+`discards`.
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:57]: How will this relate to the existing `errors` output from the remap transform?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
...
+We will also add a new command-line option and environment variable to opt into the stricter
+validation for users that want the additional assurance this provides.
+
+#### Simplify Discarded Output Handling
+
+This enforced handling will create considerable complication for most users, as well as the
+potential of increased overhead and reduced performance if extra components have to be configured
+just to handle the outputs that now need to be connected somewhere. To reduce this overhead, we will
+add a new optional configuration setting to indicate the internal disposition of each output:
+`outputs.NAME.disposition`. This will have two non-default values, `"drop"` to mark all events going
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:79]: Do you see this config applying to other outputs in the future? It sounds like we're only talking about a single hardcoded `discards` output, so I'm not sure that we need to nest the configuration in such a way that implies it could be applied to other outputs as well.

What's the benefit of `outputs.NAME.disposition` vs something like `on_discard = {send,drop,reject}`?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 9]
```diff
@@ -0,0 +1,369 @@
...
+type = "remap"
+…
+outputs.discards.disposition = "reject"
+```
+
+### Implementation
+
+#### Configuration
+
+The new options will show up in all of the enclosing configuration option structures:
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:98]: If we put them on the `Outer` types, they'll be available whether or not the inner component actually discards events. Is that what we want, or should the config only apply to components that discard?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 10]
```diff
@@ -0,0 +1,369 @@
...
+    /// Default behavior, send the event on to further components that name this as an input.
+    Send,
+    /// Discard the event and mark it is as delivered.
+    Drop,
+    /// Discard the event and mark it as failed.
+    Reject,
+}
+```
+
+#### Sources
```
Comment (lukesteensen) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:137]: Which of the possible discard scenarios apply to sources? It's not obvious that any do from the list above.

**Review by tobz — COMMENTED** [review]
This is looking good so far, but I had a few clarifying questions to make sure I understand the scope.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 11]
```diff
@@ -0,0 +1,369 @@
...
+## Scope
+
+### In scope
+
+- Handling of discarded and errored events in sources, transforms, and sinks.
+
+### Out of scope
+
+- Handling of discarded data in sources before it is translated into events.
+- Handling of discarded sink batch buffers.
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:24]: The wire-level data created by sinks out of the incoming events. This RFC is focused only on `Event` data. I will reword to clarify.

[the remark this answers is not in the document]
Comment (tobz) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:24]: Ha, looks like Luke already inquired about this one. Will resolve since you addressed it already.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 5 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:38]: Maybe, at the end of a graceful shutdown where components are being forcefully terminated. _I think_ we could still send the data to the discard output in hopes it is still connected, and/or mark them as failed.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 8 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:79]: We have some sources, like `datadog_agent` IIRC, that already have multiple outputs. If users do not want to wire up one of those outputs _and_ we enforce that all outputs be handled, then now they need a blackhole for that output as well because we special-cased the discard handling. With the generic handling, all cases are covered.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 7 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:57]: I'm not exactly sure. I am of the mind that it should be split between errors and discards (ie `abort`), but that introduces a significant breaking change. If, on the other hand, we call the new output `errors`, this works for remap but not everything that is discarded is an error.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 9 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:98]: I had hoped to avoid having another situation with that like we do for acknowledgements with the `fn can_acknowledge` in the trait. I suppose since some components will not have any output (ex `blackhole` sink), we will have to though. I will update this.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 10 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:137]: Hmm, I missed that. I was fairly certain some could error or otherwise discard after creating events (other than on output disconnect), but I'll have to review the code to see that.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 12]
```diff
@@ -0,0 +1,369 @@
...
+    disposition: OutputDisposition,
+}
+
+enum OutputDisposition {
+    /// Default behavior, send the event on to further components that name this as an input.
+    Send,
+    /// Discard the event and mark it is as delivered.
+    Drop,
+    /// Discard the event and mark it as failed.
+    Reject,
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:133]: That will be something caught by the configuration verification stage, that "handled" outputs are not named as an input. This is an issue I discussed later in the document in "Alternatives" as not having a great solution IMO.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 13]
```diff
@@ -0,0 +1,369 @@
...
+}
+```
+
+#### Sources
+
+Sources already have facilities to handle a map of named outputs in `SourceSender`. The standard
+builder will be extended to add a new standard output to accept discarded events. The `Inner` will be
+rewritten to handle either discarding events or passing them on to the sender, and the send
+functions will be extended to take the appropriate action based on this disposition.
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:142]: That's what I was thinking, to make it easier on developers to be able to say `output.discard(event);`.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 10 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:137]: Okay, AFAICT there is _exactly one_ source that can discard events between creating them and sending them to the output: `kubernetes_logs` loads events from a file source, creates events, and then runs a transform on them to parse data out, which may discard the created events if the contents don't parse properly. So, I *think* this still applies to sources because of that.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 14]
```diff
@@ -0,0 +1,369 @@
...
+are components that already have multiple outputs, or at least the option of having multiple
+outputs. Once we enforce that all outputs are handled, these will also become an issue for some
+users, which recommends the need for a more generic solution.
+
+### Unhandled Output Configuration
+
+Simplifying the configuration of previously unhandled outputs presents a bit of a Catch-22. Without
+any help from Vector, this would require users to explicitly route the unhandled outputs to a new
+`blackhole` sink. So, we want to add a shorthand to avoid the extra configuration that would
+require, and potentially the extra running component internally.
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:265]: The "without any help from Vector" is trying to say that we are _not_ helping sinks by allowing a source to configure its output to drop or reject. I will try to reword to clarify.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 15]
```diff
@@ -0,0 +1,369 @@
...
+this way. It’s also not clear how much of a performance hit this would involve, and changing how
+memory buffers work would invoke that performance penalty between nearly every single component in
+every Vector configuration. It’s still possible that some level of conditional integration with
+buffers would be a good approach, but that path towards that is less clear than solving the problem
+in isolation.
+
+## Outstanding Questions
+
+- Do we want to automate the process of rewriting configurations that have missing output handlers,
+  or can we just leave the behavior as producing an error?
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:333]: The latter. Will clarify.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 16]
```diff
@@ -0,0 +1,369 @@
...
+
+- Should the option for turning unhandled output enforcement into a hard error be written as a more
+  generic switch to turn _all_ deprecations into errors?
+
+- Do we need separate outputs to distinguish between discards and errors?
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:338]: I was thinking more of the `remap` transform, but I see it already has a `dropped` output. Perhaps then this is really all about errored events.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 17]
```diff
@@ -0,0 +1,369 @@
...
+
+## Pain
+
+Vector has a number of ways in which components may drop events due that are outside of the
+operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
```
[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:34]: Is this accurate? According to [the docs for `trait Message`](https://docs.rs/prost/latest/prost/trait.Message.html) that signature is `fn encode<B: BufMut>(&self, buf: &mut B)`.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 7 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:57]: Ah, nevermind, the `remap` transform already has a `dropped` output that is what we need, so I think all the rest of this RFC is about errors in particular and not generic "discards".

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 17 above)
[the remark this answers is not in the document]
Comment (tobz) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:34]: Hmmm.... I'm not sure why I was originally thinking it was a limitation on the `prost` side. I guess `Encodable::encode` takes `self` because we need to consume it for `proto::EventArray::from`.

The comment about not being currently applicable still applies, but at least easier to solve, potentially, than changing a trait upstream.

Good eye.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 5 above)
Comment (lukesteensen) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:35]: If we're already forcefully terminating components then it seems unlikely that we'd be able to successfully send the events anywhere. I ask mostly because this would be a failure case that'd apply to every source and transform, even those that may not have any other way of dropping event, and I'd rather focus this feature around the components where there's a more meaningful type of failure possible.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 8 above)
Comment (lukesteensen) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:77]: That's a good point, I didn't consider the effect on those sources. I am still a little bit hesitant to have configuration that refers to a named output that won't actually exist (when configured to drop or reject), since that could result in some confusion. I'm not sure I have a great alternative though, aside from introducing separate config on those sources that controls which events types are emitted.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 1 above)
Comment (jszwedko) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:70]: An alternative name for `disposition` could be `on_error` (or `on_discard`).

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 18]
```diff
@@ -0,0 +1,385 @@
...
+```toml
+[transforms.example]
+type = "remap"
+…
+outputs.errors.disposition = "reject"
+```
+
+### Implementation
+
+#### Configuration
```
Comment (jszwedko) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:94]: I wonder if we could have global configuration for the disposition as well? That might simplify it for simple use-cases. Like setting `disposition = "drop"` at the global level and have it apply to all components.

Could also do the same at the component level for those that already have multiple outputs in addition to letting them configure the disposition per-output.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 19]
```diff
@@ -0,0 +1,385 @@
...
+side and reattach the original finalizers and drop the copied events, making this process much more
+complex than for sinks, where the events don't have another output.
+
+### Sink Wrapper
+
+Instead of a wrapper running in front of sinks, we could pass an output buffer into sinks, much like
+transforms, and simply modify them to write failed events to that output. This would have the
+benefit of higher performance, as we could likely rework most existing sinks to avoid needing the
+clone and would not need the extra finalizer and task to handle forwarding the events. On the other
+hand, this performance loss appears to be relatively minor, likely in the low single digit percent
```
Comment (jszwedko) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:330]: This seems like another area that could benefit from having a copy-on-write data structure for events to avoid a full clone (I know we do have this optimization now for when the event is completely unchanged, which should help here too).

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 20]
```diff
@@ -0,0 +1,385 @@
...
+memory buffers do not work this way and it would be a significant amount of work to make them work
+this way. It’s also not clear how much of a performance hit this would involve, and changing how
+memory buffers work would invoke that performance penalty between nearly every single component in
+every Vector configuration. It’s still possible that some level of conditional integration with
+buffers would be a good approach, but that path towards that is less clear than solving the problem
+in isolation.
+
+## Outstanding Questions
+
+- Do we want to automate the process of rewriting configurations that have missing output handlers,
```
Comment (jszwedko) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:348]: I think my suggestion above of allowing global configuration of the disposition would allow users to slowly take advantage of this new behavior component-by-component.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 4 above)
Comment (jszwedko) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:353]: Hmm, yeah, the relationship between this and the discarded event metric will require some thought.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 21]
```diff
@@ -0,0 +1,385 @@
...
+of sinks. Disk buffers already have a reasonable amount of overlap with what we’d like to do, since
+they store copies of events until they have been positively acknowledged. The downside is that
+memory buffers do not work this way and it would be a significant amount of work to make them work
+this way. It’s also not clear how much of a performance hit this would involve, and changing how
+memory buffers work would invoke that performance penalty between nearly every single component in
+every Vector configuration. It’s still possible that some level of conditional integration with
+buffers would be a good approach, but that path towards that is less clear than solving the problem
+in isolation.
+
+## Outstanding Questions
```
Comment (jszwedko) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:346]: Are there any cross-cutting concerns between these new, conditional, error outputs and the configuration schema?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 22]
```diff
@@ -0,0 +1,385 @@
...
+
+There are three parts to the proposed solution:
+
+1. Add a new standardized output to all components that may discard events after errors.
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
```
Comment (jszwedko) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:50]: What will be the format of the events flowing to the `errors` output? Will we follow the precedent set by the `remap` transform of wrapping the event with some additional metadata about the failure? I think it'd be good to call that out explicitly in here.

**Review by jszwedko — COMMENTED** [review]
Nice write-up!

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 22 above)
Comment (jszwedko) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:50]: Maybe this should take advantage of the new metadata namespacing too? That feature will only exist for logs, but we could extend it to metrics and traces. Also will we offer any guarantees about the metadata, if we do include it? I could see users starting to route on it based on the type of failure. I'd suggest we push back on that clearly document that the metadata is opaque and just for human consumption for now.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 23]
```diff
@@ -0,0 +1,385 @@
...
+all component configuration traits:
+
+```rust
+trait SourceConfig: NamedComponent + Debug + Send + Sync {
+    /// Gets the list of outputs exposed by this source (existing).
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<config::Output>;
+
+    /// The actual user configuration of outputs.
+    /// Will be `None` if no such configuration is present on this component.
+    fn outputs_config(&self) -> Option<&OutputsConfig>;
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:107]: `outputs_config` would lead me to believe it would also manage other aspects of outputs, for example `datadog_agent`'s `multiple_outputs` option, or it's ability to disable specific datatype streams.

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 24]
```diff
@@ -0,0 +1,385 @@
...
+##### Output Buffer
+
+The output from transforms is fed into an `OutputBuffer` that is currently a newtype wrapper for a
+simple vector of event arrays. The additional output dispositions will be added as enum states:
+
+```rust
+pub enum OutputBuffer {
+    Drop,
+    Reject,
+    Store(Vec<EventArray>),
```
Comment (spencergilbert) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:216]: `Store` here is for events that are going to passed to a downstream component?

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 25]
```diff
@@ -0,0 +1,385 @@
...
+## Scope
+
+### In scope
+
+- Handling of discarded events in sources, transforms, and sinks.
+
+### Out of scope
+
+- Handling of discarded data in sources before it is translated into events.
+- Handling of discarded sink request data after it is translated from events.
```
[the remark this answers is not in the document]
Comment (jszwedko) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:24]: This is making me realize that end-to-end acks is another cross-cutting concern here.

[the remark this answers is not in the document]
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:24]: Noted.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 1 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:70]: It could be, but only as applies to the error output. The proposed disposition applies to all named outputs, including from the `datadog_agent` source and the `remap` transform.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 23 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:107]: The intent is to get at the configuration described by `outputs.*.disposition`. I was leaving it as a struct to have the same pattern as the acknowledgements config.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 24 above)
Comment (bruceg) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:216]: Correct. I'll note that.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 23 above)
Comment (spencergilbert) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:115]: Yeah, no issues with that - it _may_ be worth migrating all of the "outputs config" to this struct rather than having it spread between both the component config and the new output config, if only to have a more clear mental model for the UX.

It's not directly impacting this RFC as if we need this new config struct we need it, and we don't _need_ to move the other options I mentioned. Just wanted to raise that we seemed to have a new possible "group" of config options growing here.

**Comment by fuchsnj** [comment]
Overall I'm looking at what I assume are the two most common situations I think users would want.

1. A new user trying out things, who just wants to see Vector work. They don't care about errors and want them to just drop. Forcing them to create sinks to consume all errors is cumbersome / annoying. There should be an easy way to disable errors.

2. A more enterprise user that is concerned about events being dropped due to errors and wants to have a single sink that stores all the errors somewhere (s3 / kafka / etc).
   - What does the config look like here. If there are 50 components in a config, do they have to individually list out every single error output in the one sink? That seems very tedious (even if Vector enforces all the error outputs are used). 
   - Once Vector is running and starts getting errors collected, the first question I have is, what failed? Do we have information on which error occurred, which component it failed in?
   - Will the events be in a format where I can re-ingest the events. Lets say there was an outage / config error which caused millions of logs to be dead-lettered. The issue was fixed, and I now want to re-process all the dead-lettered events. (I assume this is out of scope, but it's good to at least think about)

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 26]
```diff
@@ -0,0 +1,394 @@
...
+the operator's control:
+
+- Transform processing
+- Sink encoding failure
+- Sink partitioning failure
+- Failure to write to disk buffer
+
+This leads to the potential for lost events and even silently dropped events. Users want to be able
+to capture discarded events at various points in the topology in order to forward them to fail-safe
+destinations for diagnosis.
```
Comment (benzaita) [inline rfcs/2022-08-25-12217-handling-discarded-events.md:46]: or even just to extract metrics from them to increase observability (for example, we provide a vector-based service to our customers and would like to be able to tell how many events were discarded per customer)

File: rfcs/2022-08-25-12217-handling-discarded-events.md [hunk 27]
```diff
@@ -0,0 +1,394 @@
...
+
+There are three parts to the proposed solution:
+
+1. Add a new standardized output to all components that may discard events after errors.
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
```
[the remark this answers is not in the document]
Comment (tobz) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:58]: Feel free to file an issue about this, but as worded, it sounds like what you're looking for are the existing component metrics such as `component_sent_events_total`, which sinks _do_ emit.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 26 above)
Comment (tobz) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:46]: We already emit telemetry to track dropped events (i.e. `component_discarded_events_total`), whereas the above is merely referring to a new mechanism to actually allow trying to _do_ something with those events before deciding to drop them entirely.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 27 above)
[the remark this answers is not in the document]
Comment (benzaita) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:58]: I think this might clarify what I meant: https://github.com/vectordotdev/vector/pull/14708#discussion_r1044210753

**Comment by jszwedko** [comment]
Note we will probably need to resolve https://github.com/vectordotdev/vector/issues/14969 as part of this work as it will make the incorrect type definitions much more common.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 25 above)
[the remark this answers is not in the document]
Comment (kushalhalder) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:32]: The situation here plays around like this: For an event, i.e, after all the retries have exhausted, will be marked as discarded. For a discarded event, the following will happen:
1. Push to DLQ with retries
2. If the write to DLQ was successful, acknowledge to the source for this event. 
3. if write to DLQ failed, drop the event and acknowledge to the source for this event. 
Is this a fair understanding?

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 4 above)
Comment (kushalhalder) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:362]: I agree with tobz here. In the data flow from source to sink, since all the components are sending an event to the next component, the final discard event will be thrown by the last component. 
I see the importance of the discarded metrics, it gives the true number of the events. Subsequent metrics from the dlq component helps us understand the lag behind discarded and writes to dlq, and the loss in these events too. We will need indicators, for example: the difference between the number of discarded events versus the discarded events written to the dlq component. These numbers need to be close to zero in a given window. 

TL;DR: we need discarded events metrics, and some more.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 3 above)
Comment (sweeneyb) [reply rfcs/2022-08-25-12217-handling-discarded-events.md]: Yes.  This feels similar to firewall/WAF products having "advisory" and "blocking" mode.  As a learner, it's nice to get things up and running and be warned of dangerous configurations.  Once I'm comfortable and depending on the product, I want a "strict" mode that more forcefully calls things out.

File: rfcs/2022-08-25-12217-handling-discarded-events.md (hunk 22 above)
Comment (sweeneyb) [reply rfcs/2022-08-25-12217-handling-discarded-events.md:58]: To validate your intuition, I would be tempted to route based on the metadata from an error.  Hopefully a specific example will help:

Kafka sinks can throw errors when the broker sees a message that's too large.  I'm finding it difficult to know the exact message size before sending, so it's hard to build a filter upstream.  I could see handling this situation by routing the error.  I might route to a console log or split the message and try again.  

At a minimum, I could also envision using the error data to know whether dropping the message is okay or if I need to trigger alerts.

**Comment by tanganellilore** [comment]
I proposed a PR as a starting point for answer to this RFC. This is the pull request: #24930 

I implemented this for Elasticsearch as I’m highly familiar with its architecture and the Logstash DLQ logic. While the community might eventually need support for HTTP, S3,Clickhouse or Splunk, I’ve introduced a generalized logic for sink outputs that wasn't there before, making those future integrations easier.

Let me know in PR if this is something good or not, I'm not rust expert so something can be do better for sure.

Thanks
